File: archive/p/python/binary_search.py
import sys


def binary_search(array_list, key):
    start = 0
    end = len(array_list)
    while start < end:
        mid = (start + end) // 2
        if array_list[mid] > key:
            end = mid
        elif array_list[mid] < key:
            start = mid + 1
        else:
            return mid
    return -1


def input_list(list_str):
    array_of_numbers = list_str
    array_of_numbers = array_of_numbers.split(",")
    array_of_numbers = [int(x) for x in array_of_numbers]
    return array_of_numbers


def exit_with_error():
    print('Usage: please provide a list of sorted integers ("1, 4, 5, 11, 12") and the integer to find ("11")')
    sys.exit(1)


if __name__ == "__main__":
    try:
        array_of_numbers = sys.argv[1]
        value_to_be_found = int(sys.argv[2])
        array_of_numbers = input_list(array_of_numbers)

        if array_of_numbers != sorted(
                array_of_numbers) or len(array_of_numbers) < 1:
            exit_with_error()
        index = binary_search(array_of_numbers, value_to_be_found)
        if index < 0:
            print(False)
        else:
            print(True)
    except (IndexError, ValueError):
        exit_with_error()

File: archive/p/php/binary-search.php
<?php
function usage()
{
    exit('Usage: please provide a list of sorted integers ("1, 4, 5, 11, 12") and the integer to find ("11")');
}

function parse_int($str_value)
{
    // Remove leading and trailing spaces
    $str_value = trim($str_value);

    // Make sure all digits
    if (preg_match("/^[+-]?\d+$/", $str_value) === FALSE)
    {
        return FALSE;
    }

    // Make sure valid integer
    if (
        filter_var(
            $str_value,
            FILTER_VALIDATE_INT,
            array(
                'options' => array(
                    'decimal' => TRUE,
                    'min_range' => PHP_INT_MIN,
                    'max_range' => PHP_INT_MAX
                )
            )
        ) === FALSE
    )
    {
        return FALSE;
    }

    return intval($str_value);
}

function parse_int_array($str_values)
{
    $str_array = explode(",", $str_values);
    $values = array();
    foreach ($str_array as $str_value)
    {
        $value = parse_int($str_value);
        if ($value === FALSE)
        {
            return FALSE;
        }

        array_push($values, $value);
    }

    return $values;
}

function check_sorted($values)
{
    for ($index = 1; $index < count($values); $index++)
    {
        if ($values[$index - 1] > $values[$index]) {
            return FALSE;
        }
    }

    return TRUE;
}

function binary_search($target, $values)
{
    $low = 0;
    $high = count($values);
    while ($low < $high)
    {
        $mid = intdiv($low + $high, 2);

        // If found it, return index
        if ($values[$mid] == $target)
        {
            return $mid;
        }

        // If too low, move lower bound
        if ($values[$mid] < $target)
        {
            $low = $mid + 1;
        }
        // Else, move upper bound
        else
        {
            $high = $mid;
        }
    }

    return -1;
}

// Exit if too few arguments
if (count($argv) < 3)
{
    usage();
}

// Parse 1st argument. Exit if invalid
$values = parse_int_array($argv[1]);
if ($values === FALSE)
{
    usage();
}

// Parse 2nd argument. Exit if invalid
$target = parse_int($argv[2]);
if ($target == FALSE)
{
    usage();
}

// Make sure list is sorted
if (!check_sorted($values))
{
    usage();
}

// Do binary search and show results
$index = binary_search($target, $values);
printf("%s\n", ($index >= 0) ? "true" : "false");
?>
